keanu-project/src/main/java/io/improbable/keanu/algorithms/variational/optimizer/nongradient/FitnessFunction.java

```diff
@@ -41,9 +43,14 @@ public MultivariateFunction fitness() {
         };
     }
 
-    public static double logOfTotalProbability(List<? extends Vertex> probabilisticVertices) {
+    public static double logOfTotalProbability(List<? extends Vertex> vertices) {
+        for (Vertex<?> v : vertices) {
+            if (!v.isProbabilistic() && v.isObserved() && !v.matchesObservation()) {
```

Comment:
As discussed above, `matchesObservation` doesn't make sense for probabilistic vertices. So if we put that on a new interface, we no longer need the check for `!isProbabilistic()`. Then there are 2 predicates left: `matchesObservation` and `isObserved`. I think it would be reasonable to change the semantics of `matchesObservation` to also check for `isObserved`. The two will always go together. We can roll them up into one method called `contradictsObservation`.

This actually crops up in a couple of other places: `BayesianNetwork` and `MetropolisHastingsStep`. See `BayesianNetwork` for the same logic done with 1 `for` loop instead of 2 - so I can do this elsewhere too.